### PR TITLE
dialects: Add LLVM InlineAsm op

### PR DESCRIPTION
**Pull request by JosseVanDelm**
This PR attempts to add the LLVM InlineAsm op, which allows to bake in inline assembly from within MLIR.

Some things I noted while making this:
* The docs mention this:
   > The InlineAsmOp mirrors the underlying LLVM semantics with a notable exception: the embedded asm_string is not    allowed to define or reference any symbol or any global variable: only the operands of the op may be read, written, or    referenced. Attempting to define or reference any symbol or any global behavior is considered undefined behavior at    this time.
   
   I'm assuming this means that this constraint is never actually checked/verified :sweat_smile: ?
* There's an additional attribute to select the inline assembly dialect (not dialect in context of MLIR, choices are AT&T or Intel). I just implemented this as an `IntegerAttr`, is that okay? It looked like `Linkage` (prior to this PR) was/is implemented differently from upstream in a similar way?

Currently I have only implemented round-trip tests.

Happy to add any other tests if required.

CC @jorendumoulin

### CODE REVIEW

**Comment by superlopuh** [comment]
Can you please add a test for the roundtrip with MLIR?

**Comment by JosseVanDelm** [comment]
Added @superlopuh !

File: xdsl/dialects/llvm.py [hunk 1]
```diff
@@ -712,6 +712,61 @@ def __init__(self, input: SSAValue | Operation, ptr_type: Attribute | None = Non
...
+        props: dict[str, Attribute] = {
+            "asm_string": StringAttr(asm_string),
+            "constraints": StringAttr(constraints),
+            "asm_dialect": IntegerAttr.from_int_and_width(asm_dialect, 64),
+        }
+        if has_side_effects:
+            props["has_side_effects"] = UnitAttr()
+        if is_align_stack:
+            props["is_align_stack"] = UnitAttr()
+
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:762]:
```suggestion
            "asm_dialect": IntegerAttr.from_int_and_width(asm_dialect, 64),
            "has_side_effects": UnitAttr() if has_side_effects else None,
            "is_align_stack": UnitAttr() if is_align_stack else None,
        }
```

Comment (superlopuh) [reply xdsl/dialects/llvm.py:762]: can then probably move the dict literal into the super init call

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/llvm.py [hunk 2]
```diff
@@ -712,6 +712,61 @@ def __init__(self, input: SSAValue | Operation, ptr_type: Attribute | None = Non
...
+    """
+    https://mlir.llvm.org/docs/Dialects/LLVM/#llvminline_asm-llvminlineasmop
+
+    To see what each field means, have a look at:
+    https://llvm.org/docs/LangRef.html#inline-assembler-expressions
+    """
+
+    name = "llvm.inline_asm"
+
+    # operands_, because is already defined?
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:726]: Can probably remove this comment

File: xdsl/dialects/llvm.py (hunk 1 above)
Comment (JosseVanDelm) [reply xdsl/dialects/llvm.py:762]: Pyright throws an error if I do that, cause props is not allowed to have a None as value in the dictionary.
I've seen some other shorthands here, but just like the poster here I prefer the original code:
https://stackoverflow.com/questions/14263872/only-add-to-a-dict-if-a-condition-is-met

File: xdsl/dialects/llvm.py (hunk 2 above)
Comment (JosseVanDelm) [reply xdsl/dialects/llvm.py]: Done!

File: xdsl/dialects/llvm.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py:761]: Wait a second, properties can't be missing, can you please check that in MLIR these are properties and not attributes by printing the generic syntax?

**Pull request merged**